File: src/automation/human_behavior.py
"""
Human Behavior Simulation - makes automation look like real user actions.
Implements Bezier curve mouse movements, realistic typing, scrolling patterns.
"""
import asyncio
import random
import math
from typing import Any
from dataclasses import dataclass

import numpy as np
from playwright.async_api import Page, Locator
from loguru import logger

from src.config import settings


@dataclass
class Point:
    """2D point."""
    x: float
    y: float


def gaussian_random(mean: float, std_dev: float) -> float:
    """Generate gaussian distributed random number."""
    return np.random.normal(mean, std_dev)


def bounded_gaussian(min_val: float, max_val: float) -> float:
    """Generate gaussian random within bounds."""
    mean = (min_val + max_val) / 2
    std_dev = (max_val - min_val) / 6  # 99.7% within bounds
    value = gaussian_random(mean, std_dev)
    return max(min_val, min(max_val, value))


class BezierCurve:
    """Bezier curve for smooth mouse movements."""

    @staticmethod
    def get_point(t: float, points: list[Point]) -> Point:
        """Get point on Bezier curve at parameter t (0-1)."""
        n = len(points) - 1
        x = 0.0
        y = 0.0

        for i, point in enumerate(points):
            # Bernstein polynomial
            coeff = math.comb(n, i) * (t ** i) * ((1 - t) ** (n - i))
            x += coeff * point.x
            y += coeff * point.y

        return Point(x, y)

    @staticmethod
    def generate_curve(
        start: Point,
        end: Point,
        num_control_points: int = 3,
        variance: float = 0.3
    ) -> list[Point]:
        """
        Generate Bezier curve control points between start and end.

        Args:
            start: Starting point
            end: Ending point
            num_control_points: Number of intermediate control points
            variance: How much control points can deviate from straight line

        Returns:
            List of control points including start and end
        """
        points = [start]

        dx = end.x - start.x
        dy = end.y - start.y
        distance = math.sqrt(dx * dx + dy * dy)

        for i in range(1, num_control_points + 1):
            t = i / (num_control_points + 1)

            # Point on straight line
            base_x = start.x + dx * t
            base_y = start.y + dy * t

            # Add perpendicular offset
            offset_magnitude = distance * variance * gaussian_random(0, 0.5)
            angle = math.atan2(dy, dx) + math.pi / 2

            control_x = base_x + offset_magnitude * math.cos(angle)
            control_y = base_y + offset_magnitude * math.sin(angle)

            points.append(Point(control_x, control_y))

        points.append(end)
        return points


class HumanMouse:
    """Human-like mouse movement simulation."""

    def __init__(self, page: Page):
        self.page = page
        self.config = settings.human_behavior.mouse
        self.current_pos = Point(0, 0)

    async def move_to(
        self,
        target: Point | Locator,
        steps: int | None = None,
        overshoot: bool = True
    ) -> Point:
        """
        Move mouse to target using Bezier curve.

        Args:
            target: Target point or Locator to move to
            steps: Number of intermediate points (auto-calculated if None)
            overshoot: Whether to overshoot and correct

        Returns:
            Final position
        """
        # Get target coordinates
        if isinstance(target, Locator):
            box = await target.bounding_box()
            if not box:
                raise ValueError("Target element not visible")

            # Click in center with some randomness
            target_x = box["x"] + box["width"] * random.uniform(0.3, 0.7)
            target_y = box["y"] + box["height"] * random.uniform(0.3, 0.7)
            target = Point(target_x, target_y)

        # Calculate distance
        dx = target.x - self.current_pos.x
        dy = target.y - self.current_pos.y
        distance = math.sqrt(dx * dx + dy * dy)

        # Auto-calculate steps based on distance
        if steps is None:
            steps = max(10, int(distance / 10))

        # Overshoot calculation
        overshoot_target = target
        if overshoot and random.random() < 0.7:  # 70% chance of overshoot
            overshoot_percent = random.uniform(
                self.config.overshoot_percent[0] / 100,
                self.config.overshoot_percent[1] / 100
            )
            overshoot_target = Point(
                target.x + dx * overshoot_percent,
                target.y + dy * overshoot_percent
            )

        # Generate Bezier curve
        num_control = random.randint(*self.config.bezier_points)
        control_points = BezierCurve.generate_curve(
            self.current_pos,
            overshoot_target if overshoot else target,
            num_control_points=num_control
        )

        # Calculate speed (variable throughout movement)
        speed = random.uniform(*self.config.speed_range)  # px/s
        base_delay = (distance / speed) / steps * 1000  # ms per step

        # Move along curve
        for i in range(1, steps + 1):
            t = i / steps

            # Ease in-out timing
            t_eased = self._ease_in_out(t)
            point = BezierCurve.get_point(t_eased, control_points)

            # Variable delay (slower at start and end)
            delay_multiplier = 1 + 0.5 * math.sin(math.pi * t)
            delay = base_delay * delay_multiplier * random.uniform(0.8, 1.2)

            await self.page.mouse.move(point.x, point.y)
            self.current_pos = point
            await asyncio.sleep(delay / 1000)

        # Correct overshoot if needed
        if overshoot and overshoot_target != target:
            await asyncio.sleep(random.uniform(50, 150) / 1000)
            await self._small_correction(target)

        return self.current_pos

    def _ease_in_out(self, t: float) -> float:
        """Ease in-out function for smooth acceleration/deceleration."""
        return t * t * (3.0 - 2.0 * t)

    async def _small_correction(self, target: Point):
        """Make small correction movement after overshoot."""
        steps = random.randint(3, 7)
        for i in range(1, steps + 1):
            t = i / steps
            x = self.current_pos.x + (target.x - self.current_pos.x) * t
            y = self.current_pos.y + (target.y - self.current_pos.y) * t

            await self.page.mouse.move(x, y)
            self.current_pos = Point(x, y)
            await asyncio.sleep(random.uniform(10, 30) / 1000)

    async def click(
        self,
        target: Point | Locator | None = None,
        button: str = "left",
        delay: float | None = None
    ):
        """
        Click with human-like behavior.

        Args:
            target: Target to click (moves there first if provided)
            button: Mouse button ('left', 'right', 'middle')
            delay: Delay between press and release (randomized if None)
        """
        if target:
            await self.move_to(target)

        # Random delay between press and release
        if delay is None:
            delay = random.uniform(50, 150)

        await self.page.mouse.down(button=button)
        await asyncio.sleep(delay / 1000)
        await self.page.mouse.up(button=button)

        logger.debug(f"Clicked at ({self.current_pos.x:.0f}, {self.current_pos.y:.0f})")

    async def double_click(self, target: Point | Locator | None = None):
        """Double click with human timing."""
        if target:
            await self.move_to(target)

        await self.click()
        await asyncio.sleep(random.uniform(80, 150) / 1000)
        await self.click()


class HumanTyping:
    """Human-like typing simulation."""

    def __init__(self, page: Page):
        self.page = page
        self.config = settings.human_behavior.typing

    async def type_text(
        self,
        text: str,
        element: Locator | None = None,
        clear_first: bool = False
    ):
        """
        Type text with human-like delays and occasional typos.

        Args:
            text: Text to type
            element: Element to type into (focuses if provided)
            clear_first: Whether to clear element first
        """
        if element:
            await element.focus()
            if clear_first:
                await self.page.keyboard.press("Control+A")
                await asyncio.sleep(random.uniform(100, 200) / 1000)
                await self.page.keyboard.press("Backspace")
                await asyncio.sleep(random.uniform(200, 400) / 1000)

        for i, char in enumerate(text):
            # Occasional typo
            if random.random() < self.config.typo_probability:
                await self._make_typo(char)
            else:
                await self.page.keyboard.type(char)

            # Variable delay between characters
            delay = bounded_gaussian(*self.config.delay_range)

            # Longer pauses after punctuation
            if char in ".!?":
                delay *= random.uniform(1.5, 2.5)
            elif char in ",;:":
                delay *= random.uniform(1.2, 1.5)
            elif char == " ":
                delay *= random.uniform(0.8, 1.3)

            # Occasional longer pause (thinking)
            if random.random() < 0.02:  # 2% chance
                delay += random.uniform(300, 800)

            await asyncio.sleep(delay / 1000)

        logger.debug(f"Typed {len(text)} characters")

    async def _make_typo(self, intended_char: str):
        """Make a typo and correct it."""
        # Type a nearby key
        typo_char = self._get_nearby_key(intended_char)
        await self.page.keyboard.type(typo_char)

        # Wait a bit (noticing the mistake)
        await asyncio.sleep(random.uniform(*self.config.typo_fix_delay) / 1000)

        # Delete the typo
        await self.page.keyboard.press("Backspace")
        await asyncio.sleep(random.uniform(50, 150) / 1000)

        # Type the correct character
        await self.page.keyboard.type(intended_char)

        logger.debug(f"Made typo: {typo_char} -> {intended_char}")

    def _get_nearby_key(self, char: str) -> str:
        """Get a key that's nearby on the keyboard."""
        keyboard_layout = {
            'q': ['w', 'a', 's'],
            'w': ['q', 'e', 's', 'a'],
            'e': ['w', 'r', 'd', 's'],
            'r': ['e', 't', 'f', 'd'],
            't': ['r', 'y', 'g', 'f'],
            'y': ['t', 'u', 'h', 'g'],
            'u': ['y', 'i', 'j', 'h'],
            'i': ['u', 'o', 'k', 'j'],
            'o': ['i', 'p', 'l', 'k'],
            'p': ['o', 'l'],
            'a': ['q', 'w', 's', 'z'],
            's': ['a', 'w', 'e', 'd', 'z', 'x'],
            'd': ['s', 'e', 'r', 'f', 'x', 'c'],
            'f': ['d', 'r', 't', 'g', 'c', 'v'],
            'g': ['f', 't', 'y', 'h', 'v', 'b'],
            'h': ['g', 'y', 'u', 'j', 'b', 'n'],
            'j': ['h', 'u', 'i', 'k', 'n', 'm'],
            'k': ['j', 'i', 'o', 'l', 'm'],
            'l': ['k', 'o', 'p'],
            'z': ['a', 's', 'x'],
            'x': ['z', 's', 'd', 'c'],
            'c': ['x', 'd', 'f', 'v'],
            'v': ['c', 'f', 'g', 'b'],
            'b': ['v', 'g', 'h', 'n'],
            'n': ['b', 'h', 'j', 'm'],
            'm': ['n', 'j', 'k'],
        }

        char_lower = char.lower()
        if char_lower in keyboard_layout:
            typo = random.choice(keyboard_layout[char_lower])
            return typo.upper() if char.isupper() else typo

        # For other characters, return the same
        return char


class HumanScroll:
    """Human-like scrolling simulation."""

    def __init__(self, page: Page):
        self.page = page
        self.config = settings.human_behavior.scrolling

    async def scroll_randomly(
        self,
        duration: float | None = None,
        direction_bias: float = 0.6  # 60% down, 40% up
    ):
        """
        Scroll the page randomly for a duration.

        Args:
            duration: Duration in seconds (random from config if None)
            direction_bias: Probability of scrolling down vs up
        """
        if duration is None:
            duration = random.uniform(*self.config.pre_post_duration)

        logger.debug(f"Starting random scroll for {duration:.1f}s")

        start_time = asyncio.get_event_loop().time()
        end_time = start_time + duration

        while asyncio.get_event_loop().time() < end_time:
            # Decide direction
            scroll_down = random.random() < direction_bias

            # Scroll amount (pixels)
            base_amount = random.randint(100, 500)
            amount = base_amount if scroll_down else -base_amount

            # Scroll with variable speed
            await self._smooth_scroll(amount)

            # Pause between scrolls (reading content)
            pause = random.uniform(0.5, 3.0)
            await asyncio.sleep(pause)

        logger.debug("Random scroll completed")

    async def _smooth_scroll(self, total_amount: int, steps: int = 5):
        """Smooth scroll by amount in multiple steps."""
        step_amount = total_amount / steps

        for _ in range(steps):
            # Add some variance to each step
            variance = step_amount * self.config.speed_variation
            actual_step = step_amount + random.uniform(-variance, variance)

            await self.page.mouse.wheel(0, actual_step)
            await asyncio.sleep(random.uniform(30, 80) / 1000)

    async def scroll_to_element(self, element: Locator):
        """Scroll to make an element visible."""
        # Get element position
        box = await element.bounding_box()
        if not box:
            return

        # Get viewport height
        viewport = self.page.viewport_size
        if not viewport:
            viewport = {"height": 800}

        # Calculate scroll needed
        element_center = box["y"] + box["height"] / 2
        viewport_center = viewport["height"] / 2

        scroll_needed = element_center - viewport_center

        # Scroll smoothly
        if abs(scroll_needed) > 50:
            await self._smooth_scroll(int(scroll_needed), steps=8)
            await asyncio.sleep(random.uniform(200, 500) / 1000)


class HumanBehavior:
    """
    Combined human behavior simulation.

    Usage:
        human = HumanBehavior(page)
        await human.mouse.move_to(element)
        await human.mouse.click()
        await human.typing.type_text("Hello world")
        await human.scroll.scroll_randomly(duration=30)
    """

    def __init__(self, page: Page):
        self.page = page
        self.mouse = HumanMouse(page)
        self.typing = HumanTyping(page)
        self.scroll = HumanScroll(page)

    async def random_engagement(
        self,
        duration: float | None = None,
        click_probability: float = 0.3,
        scroll_probability: float = 0.7
    ):
        """
        Perform random human-like engagement actions.

        Args:
            duration: Duration in seconds
            click_probability: Probability of clicking something
            scroll_probability: Probability of scrolling
        """
        if duration is None:
            duration = random.uniform(
                *settings.human_behavior.group_engagement.time_on_group
            )

        logger.info(f"Random engagement for {duration:.1f}s")

        start_time = asyncio.get_event_loop().time()
        end_time = start_time + duration

        while asyncio.get_event_loop().time() < end_time:
            action = random.random()

            if action < scroll_probability:
                # Scroll a bit
                scroll_duration = random.uniform(2, 8)
                await self.scroll.scroll_randomly(duration=scroll_duration)

            elif action < scroll_probability + click_probability:
                # Try to click something random
                try:
                    # Find clickable elements
                    links = await self.page.query_selector_all("a[href]")
                    if links:
                        element = random.choice(links[:10])  # Limit to first 10
                        box = await element.bounding_box()
                        if box:
                            await self.mouse.move_to(Point(
                                box["x"] + box["width"] / 2,
                                box["y"] + box["height"] / 2
                            ))
                            # Don't always click after moving
                            if random.random() < 0.3:
                                await self.mouse.click()
                except Exception as e:
                    logger.debug(f"Click action failed: {e}")

            # Random pause
            await asyncio.sleep(random.uniform(1, 4))

        logger.info("Random engagement completed")

    async def wait_human(self, min_seconds: float, max_seconds: float):
        """Wait a random human-like duration."""
        duration = bounded_gaussian(min_seconds, max_seconds)
        logger.debug(f"Human wait: {duration:.1f}s")
        await asyncio.sleep(duration)

File: nodejs-version/src/utils/human-behavior.js
/**
 * Human Behavior Simulation - Anti-Ban Stack 2025
 *
 * Implementuje:
 * - Bezier curves dla ruchu myszki (nie liniowe!)
 * - Realistyczne pisanie z opóźnieniami 120-380ms i literówkami
 * - Scrollowanie z losowymi pauzami
 * - Gaussian random zamiast uniform
 */

const { USER_AGENTS } = require('./fingerprint-manager');

// === GAUSSIAN RANDOM ===
function gaussianRandom(mean, stdDev) {
  let u = 0, v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  const num = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
  return num * stdDev + mean;
}

function boundedGaussian(min, max) {
  const mean = (min + max) / 2;
  const stdDev = (max - min) / 6;
  let value = gaussianRandom(mean, stdDev);
  return Math.max(min, Math.min(max, value));
}

// === RANDOM DELAY ===
function randomDelay(min, max) {
  // Użyj gaussian dla bardziej naturalnej dystrybucji
  const delay = boundedGaussian(min, max);
  return new Promise(resolve => setTimeout(resolve, Math.floor(delay)));
}

// === BEZIER CURVES DLA MYSZKI ===

class BezierCurve {
  static getPoint(t, points) {
    const n = points.length - 1;
    let x = 0, y = 0;

    for (let i = 0; i <= n; i++) {
      const coeff = this.binomial(n, i) * Math.pow(t, i) * Math.pow(1 - t, n - i);
      x += coeff * points[i].x;
      y += coeff * points[i].y;
    }

    return { x, y };
  }

  static binomial(n, k) {
    let result = 1;
    for (let i = 1; i <= k; i++) {
      result = result * (n - i + 1) / i;
    }
    return result;
  }

  static generateCurve(start, end, numControlPoints = 3, variance = 0.3) {
    const points = [start];
    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const distance = Math.sqrt(dx * dx + dy * dy);

    for (let i = 1; i <= numControlPoints; i++) {
      const t = i / (numControlPoints + 1);

      // Punkt na prostej linii
      const baseX = start.x + dx * t;
      const baseY = start.y + dy * t;

      // Dodaj prostopadłe przesunięcie (szum)
      const offsetMagnitude = distance * variance * gaussianRandom(0, 0.5);
      const angle = Math.atan2(dy, dx) + Math.PI / 2;

      const controlX = baseX + offsetMagnitude * Math.cos(angle);
      const controlY = baseY + offsetMagnitude * Math.sin(angle);

      points.push({ x: controlX, y: controlY });
    }

    points.push(end);
    return points;
  }
}

// === HUMAN MOUSE MOVEMENT ===

class HumanMouse {
  constructor(page) {
    this.page = page;
    this.currentPos = { x: 0, y: 0 };

    // Konfiguracja z checklisty
    this.config = {
      speedRange: [400, 1200], // px/s
      overshootPercent: [5, 15],
      bezierPoints: [3, 7],
    };
  }

  async moveTo(target, options = {}) {
    const { steps = null, overshoot = true } = options;

    // Oblicz dystans
    const dx = target.x - this.currentPos.x;
    const dy = target.y - this.currentPos.y;
    const distance = Math.sqrt(dx * dx + dy * dy);

    // Auto-oblicz liczbę kroków na podstawie dystansu
    const numSteps = steps || Math.max(10, Math.floor(distance / 10));

    // Overshoot
    let overshootTarget = target;
    if (overshoot && Math.random() < 0.7) {
      const overshootPct = boundedGaussian(
        this.config.overshootPercent[0] / 100,
        this.config.overshootPercent[1] / 100
      );
      overshootTarget = {
        x: target.x + dx * overshootPct,
        y: target.y + dy * overshootPct,
      };
    }

    // Generuj krzywą Beziera
    const numControl = Math.floor(boundedGaussian(...this.config.bezierPoints));
    const controlPoints = BezierCurve.generateCurve(
      this.currentPos,
      overshoot ? overshootTarget : target,
      numControl
    );

    // Oblicz prędkość
    const speed = boundedGaussian(...this.config.speedRange);
    const baseDelay = (distance / speed) / numSteps * 1000;

    // Ruch wzdłuż krzywej
    for (let i = 1; i <= numSteps; i++) {
      const t = i / numSteps;

      // Ease in-out
      const tEased = t * t * (3.0 - 2.0 * t);
      const point = BezierCurve.getPoint(tEased, controlPoints);

      // Zmienny delay (wolniej na początku i końcu)
      const delayMultiplier = 1 + 0.5 * Math.sin(Math.PI * t);
      const delay = baseDelay * delayMultiplier * boundedGaussian(0.8, 1.2);

      await this.page.mouse.move(point.x, point.y);
      this.currentPos = point;
      await new Promise(resolve => setTimeout(resolve, delay));
    }

    // Korekta overshoot
    if (overshoot && overshootTarget !== target) {
      await randomDelay(50, 150);
      await this.smallCorrection(target);
    }

    return this.currentPos;
  }

  async smallCorrection(target) {
    const steps = Math.floor(boundedGaussian(3, 7));
    for (let i = 1; i <= steps; i++) {
      const t = i / steps;
      const x = this.currentPos.x + (target.x - this.currentPos.x) * t;
      const y = this.currentPos.y + (target.y - this.currentPos.y) * t;

      await this.page.mouse.move(x, y);
      this.currentPos = { x, y };
      await randomDelay(10, 30);
    }
  }

  async click(target = null, button = 'left') {
    if (target) {
      await this.moveTo(target);
    }

    // Losowy delay między press i release
    const delay = boundedGaussian(50, 150);

    await this.page.mouse.down({ button });
    await new Promise(resolve => setTimeout(resolve, delay));
    await this.page.mouse.up({ button });
  }

  async clickElement(element) {
    const box = await element.boundingBox();
    if (!box) throw new Error('Element not visible');

    // Kliknij w środek z losowym przesunięciem
    const target = {
      x: box.x + box.width * boundedGaussian(0.3, 0.7),
      y: box.y + box.height * boundedGaussian(0.3, 0.7),
    };

    await this.click(target);
  }
}

// === HUMAN TYPING ===

class HumanTyping {
  constructor(page) {
    this.page = page;

    // Konfiguracja z checklisty: 120-380ms między znakami
    this.config = {
      delayRange: [120, 380],
      typoProbability: 0.03, // 3% szans na literówkę
      typoFixDelay: [200, 600],
    };
  }

  async typeText(text, element = null) {
    if (element) {
      await element.focus();
    }

    for (let i = 0; i < text.length; i++) {
      const char = text[i];

      // Losowa literówka
      if (Math.random() < this.config.typoProbability) {
        await this.makeTypo(char);
      } else {
        await this.page.keyboard.type(char);
      }

      // Zmienny delay między znakami
      let delay = boundedGaussian(...this.config.delayRange);

      // Dłuższe pauzy po interpunkcji
      if ('.!?'.includes(char)) {
        delay *= boundedGaussian(1.5, 2.5);
      } else if (',;:'.includes(char)) {
        delay *= boundedGaussian(1.2, 1.5);
      } else if (char === ' ') {
        delay *= boundedGaussian(0.8, 1.3);
      }

      // Losowa dłuższa pauza (myślenie)
      if (Math.random() < 0.02) {
        delay += boundedGaussian(300, 800);
      }

      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }

  async makeTypo(intendedChar) {
    // Wpisz literówkę
    const typoChar = this.getNearbyKey(intendedChar);
    await this.page.keyboard.type(typoChar);

    // Czekaj (zauważenie błędu)
    await randomDelay(...this.config.typoFixDelay);

    // Skasuj
    await this.page.keyboard.press('Backspace');
    await randomDelay(50, 150);

    // Wpisz poprawny znak
    await this.page.keyboard.type(intendedChar);
  }

  getNearbyKey(char) {
    const keyboard = {
      'q': ['w', 'a', 's'], 'w': ['q', 'e', 's', 'a'], 'e': ['w', 'r', 'd', 's'],
      'r': ['e', 't', 'f', 'd'], 't': ['r', 'y', 'g', 'f'], 'y': ['t', 'u', 'h', 'g'],
      'u': ['y', 'i', 'j', 'h'], 'i': ['u', 'o', 'k', 'j'], 'o': ['i', 'p', 'l', 'k'],
      'p': ['o', 'l'], 'a': ['q', 'w', 's', 'z'], 's': ['a', 'w', 'e', 'd', 'z', 'x'],
      'd': ['s', 'e', 'r', 'f', 'x', 'c'], 'f': ['d', 'r', 't', 'g', 'c', 'v'],
      'g': ['f', 't', 'y', 'h', 'v', 'b'], 'h': ['g', 'y', 'u', 'j', 'b', 'n'],
      'j': ['h', 'u', 'i', 'k', 'n', 'm'], 'k': ['j', 'i', 'o', 'l', 'm'],
      'l': ['k', 'o', 'p'], 'z': ['a', 's', 'x'], 'x': ['z', 's', 'd', 'c'],
      'c': ['x', 'd', 'f', 'v'], 'v': ['c', 'f', 'g', 'b'], 'b': ['v', 'g', 'h', 'n'],
      'n': ['b', 'h', 'j', 'm'], 'm': ['n', 'j', 'k'],
    };

    const lower = char.toLowerCase();
    if (keyboard[lower]) {
      const typo = keyboard[lower][Math.floor(Math.random() * keyboard[lower].length)];
      return char === char.toUpperCase() ? typo.toUpperCase() : typo;
    }
    return char;
  }
}

// === HUMAN SCROLLING ===

class HumanScroll {
  constructor(page) {
    this.page = page;

    // Konfiguracja z checklisty
    this.config = {
      prePostDuration: [15, 60], // sekundy przed napisaniem posta
      postPublishDuration: [45, 240], // sekundy po opublikowaniu
      speedVariation: 0.3,
    };
  }

  async scrollRandomly(duration = null, directionBias = 0.6) {
    if (duration === null) {
      duration = boundedGaussian(...this.config.prePostDuration);
    }

    const startTime = Date.now();
    const endTime = startTime + duration * 1000;

    while (Date.now() < endTime) {
      // Kierunek
      const scrollDown = Math.random() < directionBias;

      // Ilość
      const baseAmount = Math.floor(boundedGaussian(100, 500));
      const amount = scrollDown ? baseAmount : -baseAmount;

      // Płynny scroll
      await this.smoothScroll(amount);

      // Pauza (czytanie)
      await randomDelay(500, 3000);
    }
  }

  async smoothScroll(totalAmount, steps = 5) {
    const stepAmount = totalAmount / steps;

    for (let i = 0; i < steps; i++) {
      const variance = stepAmount * this.config.speedVariation;
      const actualStep = stepAmount + boundedGaussian(-variance, variance);

      await this.page.mouse.wheel({ deltaY: actualStep });
      await randomDelay(30, 80);
    }
  }

  async scrollToElement(element) {
    const box = await element.boundingBox();
    if (!box) return;

    const viewportSize = await this.page.viewportSize();
    const elementCenter = box.y + box.height / 2;
    const viewportCenter = viewportSize.height / 2;
    const scrollNeeded = elementCenter - viewportCenter;

    if (Math.abs(scrollNeeded) > 50) {
      await this.smoothScroll(scrollNeeded, 8);
      await randomDelay(200, 500);
    }
  }
}

// === LEGACY FUNCTIONS (dla kompatybilności) ===

async function randomTyping(page, selector, text, isContentEditable = false) {
  try {
    await page.waitForSelector(selector, { timeout: 5000 });
  } catch (e) {
    // Kontynuuj mimo braku selektora
  }
  await page.click(selector);
  await randomDelay(300, 600);

  const typing = new HumanTyping(page);

  // Dla kompatybilności - używamy prostszego pisania
  for (let char of text) {
    await page.keyboard.type(char);
    const charDelay = boundedGaussian(80, 200);
    await randomDelay(charDelay, charDelay + 20);

    if (Math.random() < 0.05) {
      await randomDelay(500, 1500);
    }
  }
}

async function naturalMouseMovement(page, x, y) {
  const mouse = new HumanMouse(page);
  await mouse.moveTo({ x, y });
}

async function naturalScroll(page, distance = 300) {
  const scroll = new HumanScroll(page);
  await scroll.smoothScroll(distance);
}

function getRandomUserAgent() {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
}

function calculateReadingTime(text) {
  const words = text.split(/\s+/).length;
  const wordsPerMinute = 200 + Math.random() * 50;
  const minutes = words / wordsPerMinute;
  const milliseconds = minutes * 60 * 1000;
  const variance = milliseconds * 0.3;
  return milliseconds + (Math.random() * variance * 2 - variance);
}

async function detectCaptcha(page) {
  const captchaSelectors = [
    'iframe[src*="captcha"]',
    'div[id*="captcha"]',
    '.g-recaptcha',
    'iframe[src*="recaptcha"]'
  ];

  for (const selector of captchaSelectors) {
    try {
      const element = await page.$(selector);
      if (element) return true;
    } catch (e) {
      continue;
    }
  }
  return false;
}

async function randomBreak() {
  if (Math.random() < 0.05) {
    const breakTime = boundedGaussian(10000, 40000);
    await randomDelay(breakTime, breakTime + 5000);
  }
}

async function simulateActivity(page) {
  const scroll = new HumanScroll(page);
  const mouse = new HumanMouse(page);

  const actions = Math.floor(boundedGaussian(1, 4));

  for (let i = 0; i < actions; i++) {
    if (Math.random() < 0.6) {
      await scroll.smoothScroll(boundedGaussian(100, 400));
    } else {
      const x = Math.floor(boundedGaussian(100, 900));
      const y = Math.floor(boundedGaussian(100, 700));
      await mouse.moveTo({ x, y });
    }
    await randomDelay(300, 1000);
  }
}

// === NOWE FUNKCJE DLA CHECKLISTY ===

// Engagement na grupie (20-90 sekund, 2-5 kliknięć, 0-2 lajki)
async function engageWithGroup(page, duration = null) {
  const scroll = new HumanScroll(page);
  const mouse = new HumanMouse(page);

  if (duration === null) {
    duration = boundedGaussian(20, 90);
  }

  const startTime = Date.now();
  const endTime = startTime + duration * 1000;

  let clicks = 0;
  let likes = 0;
  const targetClicks = Math.floor(boundedGaussian(2, 5));
  const targetLikes = Math.floor(boundedGaussian(0, 2));

  while (Date.now() < endTime) {
    const action = Math.random();

    if (action < 0.6) {
      // Scroll
      await scroll.scrollRandomly(boundedGaussian(2, 8));
    } else if (clicks < targetClicks) {
      // Kliknij post (bez nawigacji)
      clicks++;
      await randomDelay(1000, 3000);
    }

    await randomDelay(1000, 3000);
  }

  return { clicks, likes };
}

// Post-publish engagement (45-240 sekund)
async function postPublishEngagement(page) {
  const duration = boundedGaussian(45, 240);
  await engageWithGroup(page, duration);
}

// "Ludzki błąd" - raz na 10-15 postów
async function performHumanError(page) {
  const errorType = ['wrong_click', 'scroll_away', 'back_navigation'][Math.floor(Math.random() * 3)];

  const scroll = new HumanScroll(page);

  if (errorType === 'wrong_click') {
    await scroll.scrollRandomly(boundedGaussian(2, 5));
    await randomDelay(1000, 3000);
  } else if (errorType === 'scroll_away') {
    await scroll.smoothScroll(boundedGaussian(500, 1000));
    await randomDelay(2000, 4000);
    await scroll.smoothScroll(-boundedGaussian(500, 1000));
  } else if (errorType === 'back_navigation') {
    await page.goBack();
    await randomDelay(1000, 3000);
    await page.goForward();
    await randomDelay(1000, 2000);
  }

  return errorType;
}

module.exports = {
  // Nowe klasy
  HumanMouse,
  HumanTyping,
  HumanScroll,
  BezierCurve,

  // Funkcje pomocnicze
  gaussianRandom,
  boundedGaussian,
  randomDelay,

  // Legacy (dla kompatybilności)
  randomTyping,
  naturalMouseMovement,
  naturalScroll,
  getRandomUserAgent,
  calculateReadingTime,
  detectCaptcha,
  randomBreak,
  simulateActivity,

  // Nowe funkcje
  engageWithGroup,
  postPublishEngagement,
  performHumanError,
};
